### PR TITLE
Adding a basic setup for Cypress tests, with more tests to come

### PR DESCRIPTION
**Pull request by ironfroggy**
# Ansible Hub UI Testing

These are the integration tests for Ansible Hub UI. Please run the tests before commiting or merging
new UI functionality, or before changes to existing functionality.

## Setup the Tests

Simple install the requisite test dependencies. These are separate from the project's own dependencies:

    npm ci

## Run the Tests

The tests can be run in two modes: CLI or GUI. To run the tests in headless CLI mode:

    node_modules/.bin/cypress open

To run the tests in GUI mode, which opens the Cypress interface and an embedded browser in which to
observe and inspect test runs:

    node_modules/.bin/cypress run

## Learn more

See Cypress documenation:
    https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests.html#Folder-Structure

### CODE REVIEW

File: test/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Ansible Hub UI Testing
+
+These are the integration tests for Ansible Hub UI. Please run the tests before commiting or merging
+new UI functionality, or before changes to existing functionality.
+
+## Setup the Tests
+
+Simple install the requisite test dependencies. These are separate from the project's own dependencies:
```
Comment (ZitaNemeckova) [inline test/README.md:8]: It would be good to add that it should be run from `automation-hub-ui\test` and not `auttomation-hub-ui`. It took me a moment to realise that :)

File: test/cypress.env.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+    "host": "192.168.1.200",
```
Comment (ZitaNemeckova) [inline test/cypress.env.json:2]: Is it possible to change it to `localhost:8002`? I cannot access that IP and it makes more sense to test new code locally (for me).

File: test/cypress/integration/user_dashboard.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            cy.contains('.body', "Test F").not()
+        })
+
+        it('Can edit users', () => {
+            var username = "test"
+
+            cy.contains('.pf-c-page__header-tools .pf-c-dropdown', 'admin').click()
+            cy.contains('Logout').click()
+
+            cy.visit('https://192.168.1.200')
```
Comment (ZitaNemeckova) [inline test/cypress/integration/user_dashboard.js:91]: This should be `cy.visit(host)`

**Review by newswangerd — COMMENTED** [review]
What's `test/cypress/videos/user_dashboard.js.mp4` used for?

File: test/README.md (hunk 1 above)
Comment (ironfroggy) [reply test/README.md:8]: will do

File: test/cypress.env.json (hunk 2 above)
Comment (ironfroggy) [reply test/cypress.env.json:2]: yes, actually, this file is not supposed to be checked in and i was supposed to document that you create a local one for env-specific configuration. I'll do this.

File: test/cypress/integration/user_dashboard.js (hunk 3 above)
Comment (ironfroggy) [reply test/cypress/integration/user_dashboard.js:91]: yes it should, thank you

**Comment by ironfroggy** [comment]
@newswangerd I missed additions to .gitignore i need, that shouldn't be checked in, but when you run a test with cypress it will produce a recording of the run for inspection, by default. I'll remove this mp4 and add the appropriate gitignore line.

I'll document these generated videos, too, so people know to look for them

File: test/cypress.env.json (hunk 2 above)
Comment (ZitaNemeckova) [reply test/cypress.env.json]: @ironfroggy Can you add how to create local one? I tried to add a local config and it's not working so I had to miss something.

File: test/cypress/integration/user_dashboard.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
+var admin_menu = [
```
Comment (ZitaNemeckova) [inline test/cypress/integration/user_dashboard.js:1]: This is unused.

File: test/cypress/fixtures/users.json [hunk 5]
```diff
@@ -0,0 +1,232 @@
+[
```
Comment (ZitaNemeckova) [inline test/cypress/fixtures/users.json:1]: This seems to be unrelated to our users.

File: test/cypress/fixtures/example.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (ZitaNemeckova) [inline test/cypress/fixtures/example.json:1]: This seems extra.

File: test/cypress/fixtures/profile.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (ZitaNemeckova) [inline test/cypress/fixtures/profile.json:1]: This seems extra.

File: test/cypress/integration/user_dashboard.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    var username = Cypress.env("username")
+    var password = Cypress.env("password")
+
+    beforeEach(() => {
+        cy.visit(host)
+
+        // let e = cy.contains('div', 'Username').first()
+        // cy.log(1)
+        // Array.prototype.forEach.call(e, () => cy.log(`${p} = ${e}`))
+        // cy.log(2)
```
Comment (ZitaNemeckova) [inline test/cypress/integration/user_dashboard.js:21]: Do we need to keep this?

File: test/cypress/integration/user_dashboard.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        cy.visit(host)
+
+        cy.contains('.pf-c-form__group', 'Username').find('input').first().type(username)
+        cy.contains('.pf-c-form__group', 'Password').find('input').first().type(`${password}{enter}`)
+
+        cy.contains('#page-sidebar a', 'Users').click()
+    })
+
+    it('User table lists users', () => {
+        cy.contains('.body', "autohubtest1")
```
Comment (ZitaNemeckova) [inline test/cypress/integration/user_dashboard.js:16]: This and following line seem to be looking for non-existent users.

File: test/cypress/integration/user_dashboard.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            cy.contains('[aria-labelledby=test]', "Test F")
+            cy.contains('[aria-labelledby=test]', "Test L")
+            cy.contains('[aria-labelledby=test]', "test@example.com")
+
+            cy.contains('.body', "Test F").not()
+        })
+
+        it('Can edit users', () => {
+            var username = "test"
+
```
Comment (ZitaNemeckova) [inline test/cypress/integration/user_dashboard.js:73]: Missing `var password = "p@ssword1"`

Comment (ironfroggy) [reply test/cypress/integration/user_dashboard.js:73]: I actually decided to remove this test for now, until the API data management is working.

File: test/cypress/integration/user_dashboard.js (hunk 9 above)
Comment (ironfroggy) [reply test/cypress/integration/user_dashboard.js]: Ah, right. I was using users created by my backend tests, but we want these to be separate, so I'll remove the references.

**Comment by unlikelyzero** [comment]
Is the intent to get these running in travis eventually?

File: test/cypress.json [hunk 11]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (unlikelyzero) [inline test/cypress.json:1]: "baseUrl":"http://localhost:8002",
"reporter": "junit",
  "reporterOptions": {
    "mochaFile": "results/cypress-report-[hash].xml",
    "toConsole": true
  },

File: test/cypress/integration/user_dashboard.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
+describe('Hub User Management Tests', () => {
+    var host = Cypress.env("host")
```
Comment (unlikelyzero) [inline test/cypress/integration/user_dashboard.js:2]: You can ignore host inside tests and use baseUrl. Then you can parameterize baseUrl to work with different environment targets (qa, ci, etc)

File: test/README.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+Simple install the requisite test dependencies. These are separate from the project's own dependencies:
+
+    npm ci
+
+## Run the Tests
+
+Firsts must be run from inside the test/ directory.
+
+    cd test
```
Comment (unlikelyzero) [inline test/README.md:16]: you can skip this with a different npm script

File: test/cypress/integration/user_dashboard.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    it('User table lists users', () => {
+        cy.contains('[aria-label="User list"] [aria-labelledby=admin]', "admin")
+    })
+
+    describe("Creation and management of users", () => {
+        beforeEach(() => {
+            cy.contains('#page-sidebar a', 'Users').click()
+
+            const user = {
+                firstName: "Test F",
```
Comment (unlikelyzero) [inline test/cypress/integration/user_dashboard.js:24]: Just a FYI, there is a performance bug. The smaller the string input, the faster the tests

File: test/cypress/integration/user_dashboard.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
+describe('Hub User Management Tests', () => {
+    var host = Cypress.env("host")
+    var username = Cypress.env("username")
+    var password = Cypress.env("password")
+
+    beforeEach(() => {
+        cy.visit(host)
+
+        cy.contains('.pf-c-form__group', 'Username').find('input').first().type(username)
```
Comment (unlikelyzero) [inline test/cypress/integration/user_dashboard.js:9]: @newswangerd @ZitaNemeckova I think you'll want a better locator here

File: test/cypress/integration/user_dashboard.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
+describe('Hub User Management Tests', () => {
+    var host = Cypress.env("host")
+    var username = Cypress.env("username")
+    var password = Cypress.env("password")
+
+    beforeEach(() => {
+        cy.visit(host)
+
```
Comment (unlikelyzero) [inline test/cypress/integration/user_dashboard.js:8]: `visit` will only wait for the DOMisloaded event. If this becomes flaky, you should try to intercept the last known XHR request before the username field is rendered.

File: test/cypress/integration/user_dashboard.js [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+            cy.contains('div', 'Email').findnear('input').first().type(user.email)
+            cy.contains('div', 'Username').findnear('input').first().type(user.username)
+            cy.contains('div', 'Password').findnear('input').first().type(user.password)
+            cy.contains('div', 'Password confirmation').findnear('input').first().type(user.password)
+      
+            cy.server()
+            cy.route('POST', '/api/galaxy/v3/_ui/users/').as('post')
+            cy.contains('Save').click()
+
+            cy.contains('[aria-labelledby=test]', "Test F")
```
Comment (unlikelyzero) [inline test/cypress/integration/user_dashboard.js:42]: cy.contains('[aria-labelledby=test]', user.firstName)

File: test/cypress/integration/user_dashboard.js [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        })
+
+        afterEach(() => {
+            cy.get('.pf-c-page__header-tools .pf-c-dropdown button').click()
```
Comment (unlikelyzero) [inline test/cypress/integration/user_dashboard.js:46]: Need uniqueness or testability here

File: test/cypress/integration/user_dashboard.js [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+            cy.contains('Logout').click()
+
+            cy.contains('.pf-c-form__group', 'Username').find('input').first().type(username)
+            cy.contains('.pf-c-form__group', 'Password').find('input').first().type(`${password}{enter}`)
+    
+            cy.contains('#page-sidebar a', 'Users').click()
+
+            cy.get('[aria-labelledby=test] [aria-label=Actions]').click({"force": true})
+            cy.containsnear('[aria-labelledby=test] [aria-label=Actions]', 'Delete').click({"force": true})
```
Comment (unlikelyzero) [inline test/cypress/integration/user_dashboard.js:55]: cool!

File: test/cypress/integration/user_dashboard.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+            cy.contains('[role=dialog] button', 'Delete').click({"force": true})
+        })
+
+        it('Can create new users', () => {
+            cy.contains('[aria-labelledby=test]', "Test F")
```
Comment (unlikelyzero) [inline test/cypress/integration/user_dashboard.js:60]: is this related to a previous test? Might just throw it into a single test to avoid dependencies

File: test/cypress/integration/user_dashboard.js [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+            cy.contains('[aria-labelledby=test]', "Test L")
+            cy.contains('[aria-labelledby=test]', "test@example.com")
+
+            cy.contains('.body', "Test F").not()
```
Comment (unlikelyzero) [inline test/cypress/integration/user_dashboard.js:64]: not sure what .not is doin' hur

File: test/cypress/integration/user_dashboard.js (hunk 20 above)
Comment (unlikelyzero) [reply test/cypress/integration/user_dashboard.js:60]: usually we've used the structure `cy.contains(<locator>).should(have.value, "Test F")`
`
I don't know if it matters

**Review by unlikelyzero — APPROVED** [review]
Looks amazing! Now it'll be harder and harder to go back to python+selenium :)

I left some optional comments in line

File: test/cypress/integration/user_dashboard.js [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+            cy.contains('div', 'Password confirmation').findnear('input').first().type(user.password)
+      
+            cy.server()
+            cy.route('POST', '/api/galaxy/v3/_ui/users/').as('post')
+            cy.contains('Save').click()
+
+            cy.contains('[aria-labelledby=test]', "Test F")
+        })
+
+        afterEach(() => {
```
[the remark this answers is not in the document]
Comment (ironfroggy) [reply test/cypress/integration/user_dashboard.js:45]: it is

File: test/cypress/integration/user_dashboard.js (hunk 15 above)
Comment (unlikelyzero) [reply test/cypress/integration/user_dashboard.js:9]: oh there is a unique id already id="pf-login-username-id"

File: test/cypress.json [hunk 23]
```diff
@@ -1,4 +1,9 @@
 {
   "viewportWidth": 1200,
-  "viewportHeight": 800
+  "viewportHeight": 800,
+  "baseUrl":"http://localhost:8002",
+  "reporter": "junit",
+  "reporterOptions": {
+  "mochaFile": "results/cypress-report-[hash].xml",
+  "toConsole": true
```
Comment (ZitaNemeckova) [inline test/cypress.json:8]: @ironfroggy It's missing closing `}`. 

Fails with 
```
Error reading from: galaxy-dev/ansible-hub-ui/test/cypress.json

SyntaxError: galaxy-dev/ansible-hub-ui/test/cypress.json: Unexpected end of JSON input
```
